File: tests/integration/resource_hooks/python/step-1/__main__.py
# Copyright 2025, Pulumi Corporation.  All rights reserved.

from pulumi import (
    log,
    ResourceHook,
    ResourceHookArgs,
    ResourceHookBinding,
    ResourceOptions,
)
from random_ import Random, Component


def before_create(args: ResourceHookArgs) -> None:
    log.info(f"before_create was called with length = {args.new_inputs.get('length')}")
    assert args.name == "res", f"Expected name 'res', got {args.name}"
    assert args.type == "testprovider:index:Random", (
        f"Expected type 'testprovider:index:Random', got {args.type}"
    )


before_create_hook = ResourceHook("before_create", before_create)


def before_delete(args: ResourceHookArgs) -> None:
    log.info(f"before_delete was called with length = {args.old_inputs.get('length')}")
    assert args.name == "res", f"Expected name 'res', got {args.name}"
    assert args.type == "testprovider:index:Random", (
        f"Expected type 'testprovider:index:Random', got {args.type}"
    )


before_delete_hook = ResourceHook("before_delete", before_create)


res = Random(
    "res",
    length=10,
    opts=ResourceOptions(
        hooks=ResourceHookBinding(
            before_create=[before_create_hook],
            before_delete=[before_delete_hook],
        )
    ),
)


def fun_comp(args: ResourceHookArgs) -> None:
    childId = args.new_outputs.get("childId")
    log.info(f"fun_comp was called with child = {childId}")
    if not childId:
        raise ValueError(f"expected non empty childId, got '{childId}'")
    assert args.name == "comp", f"Expected name 'comp', got {args.name}"
    assert args.type == "testprovider:index:Component", (
        f"Expected type 'testprovider:index:Component', got {args.type}"
    )


hook_comp = ResourceHook("hook_fun_comp", fun_comp)


comp = Component(
    "comp",
    length=7,
    opts=ResourceOptions(hooks=ResourceHookBinding(after_create=[hook_comp])),
)

File: tests/integration/resource_hooks/go/step-1/main.go
package main

import (
	"errors"
	"fmt"

	"github.com/pulumi/pulumi/sdk/v3/go/pulumi"
)

func main() {
	pulumi.Run(func(ctx *pulumi.Context) error {
		beforeCreate := func(args *pulumi.ResourceHookArgs) error {
			length := int(args.NewInputs["length"].NumberValue())
			ctx.Log.Info(fmt.Sprintf("beforeCreate was called with length = %d", length), nil)
			if args.Name != "res" {
				return fmt.Errorf("expected name to be 'res', got %q", args.Name)
			}
			if string(args.Type) != "testprovider:index:Random" {
				return fmt.Errorf("expected type to be 'testprovider:index:Random', got %q", args.Type)
			}
			return nil
		}
		hookBeforeCreate, err := ctx.RegisterResourceHook("beforeCreate", beforeCreate, &pulumi.ResourceHookOptions{
			OnDryRun: false,
		})
		if err != nil {
			return err
		}

		beforeDelete := func(args *pulumi.ResourceHookArgs) error {
			length := int(args.OldInputs["length"].NumberValue())
			ctx.Log.Info(fmt.Sprintf("beforeDelete was called with length = %d", length), nil)
			if args.Name != "res" {
				return fmt.Errorf("expected name to be 'res', got %q", args.Name)
			}
			if string(args.Type) != "testprovider:index:Random" {
				return fmt.Errorf("expected type to be 'testprovider:index:Random', got %q", args.Type)
			}
			return nil
		}
		hookBeforeDelete, err := ctx.RegisterResourceHook("beforeDelete", beforeDelete, &pulumi.ResourceHookOptions{
			OnDryRun: false,
		})
		if err != nil {
			return err
		}

		res, err := NewRandom(ctx,
			"res",
			&RandomArgs{
				Length: pulumi.Int(10),
			},
			pulumi.ResourceHooks(&pulumi.ResourceHookBinding{
				BeforeCreate: []*pulumi.ResourceHook{hookBeforeCreate},
				BeforeDelete: []*pulumi.ResourceHook{hookBeforeDelete},
			}))
		if err != nil {
			return err
		}

		hookFunComp := func(args *pulumi.ResourceHookArgs) error {
			ctx.Log.Info(fmt.Sprintf("funComp was called with args = %+v", args), nil)
			childId := args.NewOutputs["childId"].StringValue()
			if childId == "" {
				return errors.New("expected non empty childId")
			}
			ctx.Log.Info(fmt.Sprintf("funComp was called with child = %d\n", childId), nil)
			if args.Name != "comp" {
				return fmt.Errorf("expected name to be 'comp', got %q", args.Name)
			}
			if string(args.Type) != "testprovider:index:Component" {
				return fmt.Errorf("expected type to be 'testprovider:index:Component', got %q", args.Type)
			}
			return nil
		}
		hookComp, err := ctx.RegisterResourceHook("hook_fun_comp", hookFunComp, &pulumi.ResourceHookOptions{
			OnDryRun: false,
		})
		if err != nil {
			return err
		}
		_, err = NewComponent(ctx, "comp", &ComponentArgs{Length: pulumi.Int(10)},
			pulumi.ResourceHooks(&pulumi.ResourceHookBinding{
				AfterCreate: []*pulumi.ResourceHook{hookComp},
			}),
		)
		if err != nil {
			return err
		}

		ctx.Export("name", res.ID())
		return nil
	})
}
